### ISSUE TITLE
NYU Minutia

### ISSUE THREAD

**Issue by A1Liu** [YACS-RCOS/yacs#308]
# NYU Course Format
NYU has several schools, and each schedules its classes slightly differently.

## General Format
Classes have a department ID, a course number, a section number, and a class number. The class number is a unique integer that can be used to identify individual classes.

#### Terms and Campuses
There are 4 academic terms: Fall, Spring, Summer, and January. The schedule is available [here][acad-cal].

There are 14 NYU Campuses:  
|           |                        |              |          |        |  
|-----------|------------------------|--------------|----------|--------|  
| Abu Dhabi | Shanghai               | Accra        | Florence | Paris  |  
| Tel Aviv  | Washington             | London       | Prague   | Sydney |  
| Madrid    | International Exchange | Buenos Aires | New York |        |  

Within the New York Campus, there are 10 schools:
- College of Arts and Sciences
- Rory Meyers College of Dentistry
- Leonard N. Stern School of Business
- Gallatin School of Individualized Study
- Liberal Studies
- Tisch School of Arts
- Tandon School of Engineering
- Steinhardt School of Culture, Education, and Human Development
- Silver School of Social Work
- School of Professional Studies

Individual classroom codes/locations can be found [here][class-codes].

#### Lectures, Labs, and Recitations
Some lectures have required labs or recitations. As far as I can tell, The course number (e.g. 120) is the same for sections of the same course regardless of whether it's a lecture or recitation. Recitations of a specific lecture are given a section number in sequence after the lecture number, so for lecture 001, the recitations could be 002,003, and 004. The section numbers follow a similar system. For a lecture with Class ID 34251, its recitations could be 34252, 34253, and 34254.

#### Waitlists
If you meet the prerequisites for a class but it's full, you can apply for its waitlist. If the class opens up, then you can then automatically enroll in the class if you so choose.

#### Multi-Departmental Classes
Some classes are identical to other classes in other departments. This is usually mentioned in the course description of the class, and really only applies to introductory classes.

#### Cross-School Classes
Depending on which school you attend, there may be limits on the amount of credits you can take outside of the school, or at least how many count towards a degree.

#### Prerequisites
Some classes require permission from the department in addition to course prerequisites. There is at least one class that has a *negative prerequisite* - that is, you need to have *not* taken a certain class to be able to register for it (MATH-UA.0235 is not open to students that have taken either MATH-UA.0233 or MATH-UA.0234).

#### Core Classes
The core requirements for a degree vary depending on which school you study at -- CAS has a larger core requirement than Tandon, for example.

#### Variable Unit Courses
Independent study courses usually have a variable amount of points.

#### Seminars and Cohort Meetings
Students at CAS and Stern have mandatory meetings with their cohorts every 2 weeks. These meetings are programmed into Albert strangely - they are each listed as their own event, but within the same course and section number. These classes are automatically registered for first year students. In addition, some students may be a part of honors programs that have their own seminars - these may or may not be available through the API, but definitely should not be visible to all end-users, as it's only relevant for a small group of students.

[acad-cal]: https://www.nyu.edu/registrar/calendars/university-academic-calendar.html
[nyu-registration-doc]: https://www.nyu.edu/students/student-information-and-resources/registration-records-and-graduation/registration.html
[class-codes]: https://www.nyu.edu/students/student-information-and-resources/registration-records-and-graduation/registration/classroom-locations.html

**Comment by Bad-Science** [comment]
@A1Liu Thanks so much for putting this together!! It's extremely helpful. I will dig into this tonight and follow up.

**Comment by A1Liu** [comment]
There's quite a few rules that technically need to be followed - I think it's not worth it to implement all of them. Probably useful to implement things like seminars and cohort meetings but things like credit limits are probably too much work to implement to be worth it. Probably best to just add a warning message for end-users that are registering for classes outside their school/too many credits.

**Comment by esilverm** [comment]
One other thing that might not have been brought up could be the fact that some schools offer dual degrees. For instance, the College of Arts and Science offers a [Dual Degree Program with Tandon](cas.nyu.edu/engineering.html). Other schools include [Gallatin](https://gallatin.nyu.edu/academics/undergraduate/universitycourses/dual-degrees.html), [Stern](http://www.stern.nyu.edu/programs-admissions/mba-programs/dual-degrees),  [Steinhardt](https://steinhardt.nyu.edu/music/admissions/masters), [School of Professsional Studies](http://www.scps.nyu.edu/academics/degree-programs/undergraduate.html), and probably some others. Some of these, such as the CAS/Tandon on have students taking classes on the Manhattan and Brooklyn campuses. I have a feeling this may not be too important but I just wanted to take note of it.

**Comment by A1Liu** [comment]
Yeah @esilverm you bring up a good point, because we want YACS to be comprehensive enough to not restrict or confuse the end user, but not so comprehensive that we spend months simply coding in edge-cases. I guess we should come up with a list of edge cases that NEED to be supported, and then decide from there which parts of the NYU system should be included, because I think dual degrees are a pretty important edge case that falls under "stuff we should cover"

**Comment by A1Liu** [comment]
### General
So classes are a single meeting time for a specific course. In the course registration system they can be a lecture, recitation, or lab

### Campuses
The campuses definitely should be integrated somehow - students at the new york campus shouldn't see classes in Abu Dhabi or London and vice versa. It's possible that it could be integrated into the initial menu of the UI, but that could be dangerous - there's quite a few different campuses and departments within each campus.

### Lectures, labs, and recitations
I don't know that the general model will work. For a given course, there is a set of available lectures each taught at a different time. For each lecture, there may be a set of labs or recitations that are each taught at different times. Thus, recitations for Lecture 1 shouldn't be available to a student that has selected lecture 2, and there should be some kind of notification when a lecture is selected without a recitation.

### Waitlists
Ok, waitlists are kind of a 'the end user should probably be on top of this' kind of thing, so it's probably like really low priority.

I don't know that a status value is even necessary. Since you control the way that the data is stored in the database, you can simply have different combinations of values mean different things -- i.e. when `seats available == 1 + seats total`, that means the class is closed. When `seats available == -1` that means 1 person is on the waitlist and so on. (also if section.status is implemented, it should be as a 'magic number' instead of a string).

### Multi-department courses
This could be done as a simple edge-list. It'd just need a single extra table in the database with 4 features: integer primary key `id`, string `uuid`, integer/string `course_id_1`, and integer/string `course_id_2`. We could also potentially include a "multidept" tag in the course tags list to reduce the amount of querying that we do to the edge-list table.

### Cross-school classes
This is probably low priority, but it does vary wildly from department to department, so ideally we'd be able to help the end user out a little. It's not really a big deal though.

### Variable unit courses
That's fine, like I think that only independent study courses even have the potential for this anyways.

### Seminars and Cohort Meetings
I don't necessarily think this should be a normal course. There are something like 30 cohorts in CAS alone, and maybe 10-15 honors scholars sections. Regardless of how it's formatted in the UI, having so many classes would likely clutter the UI for the average end-user who isn't in CAS or Stern, and scrolling through the sections wouldn't be ideal. Maybe it can be done through a simple menu? Since these classes aren't really things you can opt in or out of, we could potentially just ask the user to write in the course ID/cohort number themselves - they'd have to know it as part of the registration process.

### Prerequisites
I think this should definitely be done through an edge-list table and an entirely separate/additional UI section. Ideally, the end-user should be able to signify that they are part of a specific program as a way to narrow down the list of available courses - this is more of a pipe dream than a realistic goal, but it is something to work towards eventually. It's probably low priority for now though.

I think that in general, the philosophy of "general system" that's university agnostic is great, but I don't know that the Yacs system is flexible enough yet. Lectures and recitations for sure are features that are a high priority - they're an integral part of how many of the large lecture-based classes function, and will be a problem for 90% if not 100% of the NYU user-base. Without an integrated system to handle the lectures and recitations, NYU students will essentially be forced to simultaneously use Yacs and Albert at the same time in order to verify that the sections they have chosen in Yacs are actually compatible.

**Comment by Bad-Science** [comment]
Ok, cool, thanks for the reply.

So it seems like we can divide the issues into things that are breaking, things that aren't breaking but are important to the experience, and nice to haves/future plans. I agree we need to add some flexibility in order for this to work nicely. For our first NYU release, however, we want to have the minimum viable functionality, and can continue enhancing things as we go. I don't want to rush any design decisions that will be hard to change later.

This is what I propose:

## Breaking Issues
We absolutely need to solve these before our first NYU release

### Labs, Lectures, and Recitations
There *might* be a way to get this to fit into the existing model of sections. The closest thing we have are `periods`, which belong to sections. In our current model, every section has a list of periods(classes). These can be lectures, labs, recitations, etc. But those are fixed per section, so when you sign up for a section, it comes with its set of periods (classes), all of which are required. From what I understand based on your comments, at NYU, once you pick a section, you then have more granularity and are able to pick different lectures, recitions, labs etc within that section. **Is that correct?** If so, then we will have to change our sections model to handle both cases. I don't think it would be too problematic, since the RPI sections are sort of a special case of the NYU sections, but our first task should be figuring out the best way to model this and represent it in the UI.

## Important Issues
We should try to solve these in the short term, but they are not completely vital for our first release

### Section status/waitlists
I still think having a status column would be nice to get into the initial release. Not all schools are able to provide seat numbers, and I haven't seen a source for seat numbers at NYU yet. It's very easy to add this too. We also do not want to make assumptions about how relevant seat counts might be at other schools, and having this field frees us from doing logic around seat numbers.

### Multi-department courses
I think that's the basic idea, but it's a little more complicated than that. We will need to figure out how to represent these in the API, and how to ingest them properly and build this into the pipeline. We also need to figure out what data should be de-duplicated and where. At RPI, cross-listed courses also share sections, sort of, so we'd need to handle that. I'd put this at the bottom of the "important issues" list in terms of priority. We can design it, but I don't want to try to implement it before the first launch.

### Seminars and Cohorts
I don't think I really understand how these work still. What I do not want to do is build any kind of special UI just for this. We can discuss this at our meeting so I can fully understand what they are and how we can best deal with them without too much special care or clutter. I would also consider this something that we could not include in our initial launch if there is not an easy solution. I'd rather have a few things perfect than hack in new features, even if they are very useful.

### Campuses
I agree that navigating Yacs without this organization would be unpleasant. For our first release however, we will be testing on just one campus to start (this is for logistical reasons as well). So while I would consider this "breaking" for NYU in general, It is not necessary for our first release.

## Nice to haves/future issues
These things are not on the short-term roadmap, but we want to do them

### Prerequisites
This is a little more complicated when it sounds when you consider how we would actually ingest and store this information. I'm sure we will figure it out, but need need to get the design right. I think it's something we should work on together with school # 3 (TBD), so we will have a better idea of how prerequisites are structured at other schools. Not all prerequisites are well defined, and in fact at RPI many of them are not. We often see things like `A or (B AND C)` and `A and B or "an upper level match class"`, or things even more complex combinations and arbitrary values. In an ideal world it's an easy problem, but there are MANY edge cases that we would want to handle elegantly. But we can discuss that when the time comes.

**Comment by A1Liu** [comment]
Yeah, so at NYU you first pick a section and then afterwards if there is a recitation you pick one of a set of applicable recitations. We could probably add a drop-down menu after you pick a section that displays all the applicable periods - that way there's a visual clue that the recitations are specifically for that lecture. 

To be honest, a lot of the students don't entirely understand cohorts either - there's still a few parts that I'm not sure on.

Those categories sound good!

**Comment by A1Liu** [comment]
@Bad-Science So there's usually maybe 4 recitations per lecture, and maybe 2-3 lectures per course, but both of those ranges can vary. It' probably be ok - recitations are only ever for 1 specific lecture, so the sets of recitations that the lectures have are all mutually exclusive. It'd just be somewhat confusing - there would have to be room in the UI for 2 sets of class id's, 2 sets of instructors, etc.

**Comment by Bad-Science** [comment]
@A1Liu OH! Okay, so then once I pick a lecture, my recitations and other classes for that course are predetermined?

**Comment by A1Liu** [comment]
@Bad-Science yeah exactly. You pick a lecture section and then you usually just need to pick one recitation/lab section as well

**Comment by Bad-Science** [comment]
Ok. If I am understanding you then that makes this easier. It sounds like we may not need to change the actual data model (significantly), we may just want to make some ui refinements to make sure everything is intuitive to NYU students as it is to RPI students. I'm going to look at the API tonight and see what the data looks like exactly, and we can discuss what exact changes we need to make during our meeting and report back here. Thanks for all the help!
